Add Kubernetes Access FAQ and Troubleshooting docs (#29632)
gravitational/teleport: docs/pages/kubernetes-access/troubleshooting.mdx

---

> title: Kubernetes Access Troubleshooting Guide

Reviewer: ```suggestion
title: Kubernetes Access Troubleshooting
```

description: Troubleshooting common issues with Kubernetes access
---

> 

Reviewer: Consider adding an introduction sentence or two.

## An agent failed to join a cluster due to "no authorities for hostname"

### Symptoms

The agent can't rejoin the Teleport cluster after restart and reports an error
similar to:

```text
ssh: handshake failed: ssh: no authorities for hostname
```

### Explanation

> Teleport uses Certificate Authorities (CAs) to sign certificates for each

Reviewer: ```suggestion
Teleport uses certificate authorities (CAs) to sign certificates for each
```

component. When the component joins the cluster for the first time, it
receives a certificate signed by the cluster's CA and stores it in its
state directory. When the component restarts, it uses the certificate stored in
its state directory to join the cluster again.

This error occurs when the component tries to rejoin the Teleport cluster, but the
cluster's CA has changed and the component's certificate is no longer valid.

It can happen when the cluster's CA is rotated or when the cluster is recreated
or renamed.

### Resolution

> The agent's state needs to be reset, so it requests a new certificate and joins
> the cluster again.

Reviewer: As written this could be read as "because it needs to be reset, it requests blah blah on it's own. Suggested rephrase:

```suggestion
The agent's state needs to be reset, so it can request a new certificate and join
the cluster again.
```

on behalf of the user. This is done by sending a request to the Kubernetes API server
with the user's identity in the `Impersonate-User` header and all the Kubernetes
Groups they can use in the `Impersonate-Group` header.

Since `system:masters` is a built-in Kubernetes group in all clusters, it's usual
for administrators to use it to gain administrative access to the cluster's control
plane. However, Autopilot clusters forbid impersonating this group and require that
another group is used instead.

### Resolution

Per the description above, the solution is to configure a different group for
impersonation in Teleport. This can be done by setting the role's
`kubernetes_group` parameter to a Group that the Autopilot cluster allows to
impersonate.

The `teleport-kube-agent` chart configures a Kubernetes Group with the same access
levels as the `system:masters` group when it detects the target cluster is a GKE
cluster.
This group is named, by default, `cluster-admin`, but the name can be changed
by setting the `adminClusterRoleBinding.name` parameter.

The Kubernetes Group isn't created automatically when installing the chart on

> non-GKE clusters, so please don't change the `kubernetes_groups` parameter to

Reviewer: ```suggestion
non-GKE clusters, so don't change the `kubernetes_groups` parameter to
```

`cluster-admin` unless you created the group manually or installed the chart with
`adminClusterRoleBinding.create` parameter set to `true`.

It's important to note that the group must be configured in the cluster before
it can be used for impersonation. If the group is not configured, the impersonation
request will fail with a `403 Forbidden` error and the user will not be able to
access the cluster.

When you opt to continue using the `system:masters` group for impersonation in
non-Autopilot clusters, you must ensure that the Teleport roles that grant
`system:masters` access can't be used to access GKE Autopilot clusters.

As an example, a user with the following role can impersonate the `system:masters`
group in any Kubernetes cluster:

```yaml
kind: role

> version: v5

Reviewer 2: Should we make this a v6 role?

Author: Actually for master, it should be `v7` :laughing:

---
title: Kubernetes Access Troubleshooting
description: Troubleshooting common issues with Kubernetes access
---

> This guide describes common issues with Kubernetes Access and how to resolve them.

Reviewer: This is super nit-picky, and we don't yet have this written into our style guide, but watch that not stop me 😄 

I would call this a "doc" instead of a "guide". I see a guide as a subset of documentation that specifically 'guides' the reader through a process. This is a collection of useful information, more akin to 'reference' material, but not explicitly so. Therefore: 

```suggestion
This page describes common issues with Kubernetes Access and how to resolve them.
```
